paper/code_release/utils.py

```diff
@@ -0,0 +1,247 @@
+    # creates gninatype file for model input
+
+    types_file = Path(output_dir) / f"{Path(pdb_file).stem}.types"
+    gninatypes_file = Path(output_dir) / f"{Path(pdb_file).stem}.gninatypes"
+
+    with open(types_file, "w") as f:
+        f.write(str(pdb_file))
+
+    atom_map = molgrid.GninaIndexTyper()
+    dataloader = molgrid.ExampleProvider(atom_map, shuffle=False, default_batch_size=1)
+    dataloader.populate(str(types_file))
+    example = dataloader.next()
+    coords = example.coord_sets[0].coords.tonumpy()
+    types = example.coord_sets[0].type_index.tonumpy()
+    types = np.int_(types)
+    with open(gninatypes_file, "wb") as f:
+        for i in range(coords.shape[0]):
+            f.write(struct.pack("fffi", coords[i][0], coords[i][1], coords[i][2], types[i]))
+    os.remove(str(types_file))
+    return gninatypes_file
+
+
+def sdf_to_gninatypes(sdf_file: Union[str, Path], output_dir: Union[str, Path]) -> List[Path]:
+    sdf_file, output_dir = Path(sdf_file), Path(output_dir)
+    conformer_position_types = sdf_to_xyz_atomic(sdf_file=sdf_file)
+    if len(conformer_position_types[0]) <= 1:
+        raise RuntimeWarning(
+            f"Cannot process {sdf_file}. Molecule has too few atoms: {conformer_position_types[0]} atoms."
+        )
+
+    output_file_names = []
+    for conformer_index, conformer in enumerate(conformer_position_types):
+        output_file_name = output_dir / f"{sdf_file.stem}_{conformer_index}.gninatypes"
+        with open(output_file_name, "wb") as f:
+            for atom_coordinates in conformer:
+                if atom_coordinates.gninatype_index != -1:
+                    f.write(
+                        struct.pack(
+                            "fffi",
+                            atom_coordinates.x,
+                            atom_coordinates.y,
+                            atom_coordinates.z,
+                            atom_coordinates.gninatype_index,
+                        )
+                    )
+                else:
+                    print(f"Skipping atom coordinate {atom_coordinates.atomic_num}")
+        output_file_names.append(output_file_name)
+    return output_file_names
+
+
+def make_types_file(
+    sdf_gninatypes: List[Union[str, Path]],
+    pdb_gninatypes: Union[str, Path],
+    out_file: Union[str, Path],
+):
+    tmp = [0] * len(sdf_gninatypes)
+    types_df = pd.DataFrame(
+        dict(
+            pose=tmp,
+            affinity=tmp,
+            rmsd=tmp,
+            docking=tmp,
+            protein_gninatype=[str(pdb_gninatypes)] * len(sdf_gninatypes),
+            ligand_gninatype=sdf_gninatypes,
+        )
+    )
+    types_df.to_csv(out_file, sep=" ", index=False, header=False)
+    return types_df
+
+
+warn(
+    "We are adding Hydrogens using pybel. This might not be suitable in some conditions. Please revise this in the future."
+)
+warn("Also we are adding hydrogens but are not modelling the hydrogens in the protein... no need!")
```

Comment:
Maybe all of the warnings should be at the top of the file? Or these two seem to be specific to the `sdf_to_xyz_atomic` function, so maybe they should go inside of it?

Reply:
I suggest not touching the code for now, it is research code